occ LDAP update (#4300)
owncloud/docs: modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_ldap_integration_commands.adoc

[source,console]
----
ldap
 ldap:check-user               Checks whether a user exists on LDAP
 ldap:create-empty-config      Creates an empty LDAP configuration
 ldap:delete-config            Deletes an existing LDAP configuration
 ldap:invalidate-cache         Invalidates the LDAP cache
 ldap:search                   Executes a user or group search
 ldap:set-config               Modifies an LDAP configuration
 ldap:show-config              Shows the LDAP configuration
 ldap:test-config              Tests an LDAP configuration
----

== Search for a User

Search for an LDAP user, using this syntax:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:search [options] [--] <search>
----
=== Arguments

[width="100%",cols="20%,70%",]
|===
| `search`
| The search string (can be empty)
|===

=== Options

[width="100%",cols="20%,70%",]
|===
| `--group`
| Searches groups instead of users

| `--offset=OFFSET`
| The offset of the result set. Needs to be a multiple of limit. defaults to 0. [default: 0]

| `--limit=LIMIT`
| Limit the results. 0 means no limit, defaults to 15 [default: 15]
|===

> Searches match at the beginning of the attribute value only. This example searches for `givenNames` that start with 'rob':

Reviewer: ```suggestion
Searches match at the beginning of the attribute value only. This example searches for `givenNames` that contain 'rob':
```

[source,console]
----
ldap
 ldap:check-user               Checks whether a user exists on LDAP
 ldap:create-empty-config      Creates an empty LDAP configuration
 ldap:delete-config            Deletes an existing LDAP configuration
 ldap:invalidate-cache         Invalidates the LDAP cache
 ldap:search                   Executes a user or group search
 ldap:set-config               Modifies an LDAP configuration
 ldap:show-config              Shows the LDAP configuration
 ldap:test-config              Tests an LDAP configuration
----

== Search for a User

Search for an LDAP user, using this syntax:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:search [options] [--] <search>
----
=== Arguments

[width="100%",cols="20%,70%",]
|===
| `search`

> | The search string operates against the users fullname (`cn`). It is a by default a prefix search.

Reviewer 2: > The search string operates against the users fullname (`cn`)

I'm quite sure this is by default. It's configurable in the ldap wizard; there is a section to use search attributes in the advanced tab. In fact, it uses the configured displayname attribute, which happens to be the "cn".

It could be something like:
Search attributes -> ownCloud's displayname attribute (by default) -> "cn" (by default)

| Use the empty string `''` to list all users. Start the string with `*` to change to infix (substring) search.
|===

=== Options

[width="100%",cols="20%,70%",]
|===
| `--group`
| Searches groups instead of users

| `--offset=OFFSET`

> | The offset of the result set. Needs to be a multiple of limit. defaults to 0. [default: 0]

Reviewer 2: > Needs to be a multiple of limit

I don't think this is a real requirement. You might want this to give a sense of pagination (get the first page, then the second, then the next one...), but as said, not a real requirement.

[source,console]
----
ldap
 ldap:check-user               Checks whether a user exists on LDAP
 ldap:create-empty-config      Creates an empty LDAP configuration
 ldap:delete-config            Deletes an existing LDAP configuration
 ldap:invalidate-cache         Invalidates the LDAP cache
 ldap:search                   Executes a user or group search
 ldap:set-config               Modifies an LDAP configuration
 ldap:show-config              Shows the LDAP configuration
 ldap:test-config              Tests an LDAP configuration
----

== Search for a User

Search for an LDAP user, using this syntax:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:search [options] [--] <search>
----
=== Arguments

[width="100%",cols="20%,70%",]
|===
| `search`
| The search string (can be empty)
|===

Note that the search sting can be empty to list all users, or groups when the `--group` option is set. In this case it must be declared with `''`. It is by default a prefix search. Start the string with `*` to change to infix (substring) search.

> Search attributes are configurable in the LDAP wizard (menu:Settings[Admin > User Authentication > LDAP > Advanced > User Search Attributes]. Which are, if not otherwise defined: Search attributes -> ownCloud's displayname attribute (by default) -> "cn" (by default).

Reviewer 2: "Search attributes are configurable in the LDAP wizard (menu:Settings[Admin > User Authentication > LDAP > Advanced > User Search Attributes]). If the list of attributes is empty, ownCloud will use whatever attribute is set as ownCloud's displayname attribute (menu:Settings[Admin > User Authentication > LDAP > Advanced > User Display Name Field])
For groups, the behaviour is the same, although they have their own settings: Group Search Attributes and User Display Name Field".

I'm not sure if it's clear enough for the groups, but I didn't want to repeat the same stuff again.

Reviewer 2: Note that for users, the default displayname attribute is "displayname" not "cn". For groups, the default displayname attribute is "cn".

Author: I think I am getting closer, will file an update

Author: updated, pls have a look.
btw more small issues fixed 😄

Reviewer 2: > Search attributes are configurable in the LDAP wizard or via `ldap:set-config` as described below:
>
> .For users, the default displayname attribute is "displayname"
> menu:Settings[Admin > User Authentication > LDAP > Advanced > User Search Attributes]

It seems a paragraph is missing... maybe "If the search attributes are empty, the default displayname attributes set in the configuration will be used, otherwise only the configured search attributes will be used, in particular, the 'displayname' attribute might need to be explicitly added to the list"

This way, both paragraphs seem more connected.

Author: > attributes set in the configuration will be used

in which configuration, where are these configurations located, are the gui settings different to the occ settings? not clear to me. Not today, but we should have a call tomorrow to clarify.
We need to anyways revise the ldap documentation where we also need to address this.

Author: I have improved the search attribute description according our discussion - pls have a look

{occ-command-example-prefix} ldap:search "*rob"
----

User search attributes are set with `ldap:set-config` (below). For example, if your search attributes are `givenName` and `sn` you can find users by first name + last name very quickly. For example, you’ll find 'Terri Hanson' by searching for `te ha`. Trailing whitespace is ignored.

== Check if an LDAP User Exists

This works only if the ownCloud server is connected to an LDAP server.

[source,console,subs="attributes+"]
----
ldap:check-user [options] [--] <ocName>
----

=== Arguments

[width="100%",cols="20%,70%",]
|===
| `ocName`

> | The user name as used in ownCloud

Reviewer: ```suggestion
| The user name as used in ownCloud (see e.g. the user's Federated Cloud ID).
```

----

== Delete an Existing LDAP Configuration

Deletes an existing LDAP configuration.

[source,console,subs="attributes+"]
----
 ldap:delete-config <configID>
----

=== Arguments

[width="100%",cols="20%,70%",]
|===
| `configID`
|  The configuration ID
|===

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:delete  s01
Deleted configuration with configID 's01'
----

== Invalidate LDAP Cache

This command invalidates the LDAP cache for all users:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:invalidate-cache
----

== Manipulate LDAP Configurations

This command manipulates LDAP configurations.

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:set-config <configID> <configKey> <configValue>
----

=== Arguments

[width="100%",cols="20%,70%",]
|===
| `configID`
|  The configuration ID

| `configKey`
|  The configuration key

| `configValue`
|  The new configuration value
|===

This example sets search attributes:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} ldap:set-config s01 ldapAttributesForUserSearch "cn;givenname;sn;displayname;mail"
----

> All of the available keys, along with default values for configValue, are listed in the table below.

Reviewer: ```suggestion
Available keys, along with default values for configValue, are listed in the table below.
```

Never say all. This table has 50 entries, My `occ ldap:show-config s01 | wc -l` reports 58 (including 4 non-body lines).
